File: bindings/wasm/test/portable-parity.test.ts
import { existsSync, readFileSync } from "node:fs";
import { resolve } from "node:path";
import test from "node:test";
import assert from "node:assert/strict";

const api = await import("../src/index.ts");
const generatedModulePath = resolve(import.meta.dirname, "../pkg/prolly_wasm.js");
const generatedWasmPath = resolve(import.meta.dirname, "../pkg/prolly_wasm_bg.wasm");
const generatedPresent = existsSync(generatedModulePath) && existsSync(generatedWasmPath);
let wasm: any;
if (generatedPresent) {
  wasm = await import("../pkg/prolly_wasm.js");
  wasm.initSync({ module: readFileSync(generatedWasmPath) });
}

const bytes = (value: string): Uint8Array => new TextEncoder().encode(value);

test("WASM exposes portable maps and explicit native exclusions", { skip: !generatedPresent }, async () => {
  assert.equal(typeof api.Engine, "function");
  const engine = api.Engine.memory(wasm);
  assert.throws(() => api.Engine.file(wasm, "db.prolly"), /unsupported.*wasm/i);
  assert.throws(() => api.Engine.sqlite(wasm, "db.sqlite"), /unsupported.*wasm/i);

  const versioned = engine.versionedMap(bytes("users"));
  await versioned.initialize();
  await versioned.put(bytes("u1"), bytes("Ada"));
  assert.equal(Buffer.from(await versioned.get(bytes("u1")) ?? []).toString(), "Ada");

  const registry = engine.indexRegistry();
  registry.register({
    name: bytes("by_team"), generation: 1n, extractorId: "team-v1", projection: "all",
    extract: (_key: Uint8Array, source: Uint8Array) => [{ term: Uint8Array.from(source) }],
  });
  const indexed = engine.indexedMap(bytes("members"), registry);
  await indexed.put(bytes("u1"), bytes("red"));
  await indexed.ensureIndex(bytes("by_team"));
  const snapshot = await indexed.snapshot();
  const index = snapshot.index(bytes("by_team"));
  const rows = await index.records(bytes("red"));
  assert.equal(Buffer.from(rows[0].primaryKey).toString(), "u1");
  await indexed.put(bytes("u2"), bytes("red"));
  assert.equal((await index.records(bytes("red"))).length, 1, "snapshot remains pinned");
  const freshSnapshot = await indexed.snapshot();
  const freshIndex = freshSnapshot.index(bytes("by_team"));
  assert.equal((await freshIndex.records(bytes("red"))).length, 2);
  let escaped: Uint8Array | undefined;
  await index.exactView(bytes("red"), (row: any) => {
    escaped = row.primaryKey;
    assert.equal(Buffer.from(row.primaryKey).toString(), "u1");
    return false;
  });
  assert.throws(() => escaped?.byteLength, /expired/i);

  const proximity = await engine.buildProximity(2, [
    { key: bytes("a"), vector: new Float32Array([0, 0]), value: bytes("alpha") },
  ]);
  const session = proximity.read();
  assert.equal(session.contains(bytes("a")), true);
  assert.equal(Buffer.from(session.get(bytes("a"))?.value ?? []).toString(), "alpha");
  const result = await session.search({
    vector: new Float32Array([0.1, 0.1]), topK: 1, policy: "exact",
  });
  assert.equal(Buffer.from(result.neighbors[0].key).toString(), "a");
  assert.equal(proximity.count(), 1n);
  assert.equal(proximity.contains(bytes("a")), true);
  assert.equal(proximity.config().dimensions, 2);
  assert.equal(proximity.proveStructure().verify(proximity.descriptor()).summary.recordCount, 1n);
  const mutation = proximity.mutate([
    { key: bytes("b"), vector: new Float32Array([1, 1]), value: bytes("beta") },
  ]);
  assert.equal(mutation.map.count(), 2n);
  assert.ok(mutation.stats.recordsRebuilt >= 1n);

  session.close();
  proximity.close();
  freshIndex.close();
  freshSnapshot.close();
  index.close();
  snapshot.close();
  indexed.close();
  registry.close();
  versioned.close();
  engine.close();
});

test("WASM promise wrappers own inputs and honor AbortSignal", { skip: !generatedPresent }, async () => {
  const engine = api.Engine.memory(wasm);
  const map = engine.versionedMap(bytes("async"));
  await map.initialize();
  const key = bytes("original-key");
  const value = bytes("original-value");
  const pending = map.put(key, value);
  key.fill(0);
  value.fill(0);
  await pending;
  assert.equal(Buffer.from(await map.get(bytes("original-key")) ?? []).toString(), "original-value");

  const controller = new AbortController();
  controller.abort();
  await assert.rejects(map.get(bytes("original-key"), controller.signal), /abort/i);
  map.close();
  engine.close();
});

test("WASM versioned maps expose identity and historical snapshot lifecycle", { skip: !generatedPresent }, async () => {
  const engine = api.Engine.memory(wasm);
  const map = engine.versionedMap(bytes("versioned-lifecycle"));
  assert.equal(Buffer.from(map.id()).toString(), "versioned-lifecycle");
  assert.equal(await map.isInitialized(), false);
  const initial = await map.initialize();
  assert.equal(await map.isInitialized(), true);
  assert.deepEqual(await map.headId(), initial.id);
  const first = await map.put(bytes("k"), bytes("v1"));
  await map.put(bytes("k"), bytes("v2"));
  assert.deepEqual((await map.head())?.id, await map.headId());
  assert.deepEqual((await map.version(first.id))?.id, first.id);
  assert.ok((await map.versions()).length >= 3);
  const historical = await map.snapshotAt(first.id);
  assert.ok(historical);
  assert.deepEqual(historical.id(), first.id);
  assert.deepEqual(historical.version().id, first.id);
  assert.equal(Buffer.from(historical.get(bytes("k")) ?? []).toString(), "v1");
  engine.close();
});

test("WASM versioned maps expose owned batch CAS and version-pinned point reads", { skip: !generatedPresent }, async () => {
  const engine = api.Engine.memory(wasm);
  const map = engine.versionedMap(bytes("versioned-cas"));
  await map.initialize();
  const first = await map.apply([
    { kind: "upsert", key: bytes("a"), value: bytes("one") },
    { kind: "upsert", key: bytes("b"), value: bytes("two") },
  ]);
  assert.equal(await map.containsKey(bytes("a")), true);
  assert.deepEqual((await map.getMany([bytes("a"), bytes("missing")])).map((value: Uint8Array | undefined) => value == null ? undefined : Buffer.from(value).toString()), ["one", undefined]);
  const applied = await map.putIf(first.id, bytes("a"), bytes("updated"));
  assert.equal(applied.kind, "applied");
  const conflict = await map.deleteIf(first.id, bytes("b"));
  assert.equal(conflict.kind, "conflict");
  const values = await map.getManyAt(first.id, [bytes("a"), bytes("b")]);
  assert.deepEqual(values.map((value: Uint8Array | undefined) => Buffer.from(value ?? []).toString()), ["one", "two"]);
  assert.equal(Buffer.from(await map.getAt(first.id, bytes("a")) ?? []).toString(), "one");
  const batch = await map.applyIf(applied.current!.id, [{ kind: "delete", key: bytes("b") }]);
  assert.equal(batch.kind, "applied");
  engine.close();
});

test("WASM proofs, retained sessions, and maintenance stay in Rust", { skip: !generatedPresent }, async () => {
  const engine = api.Engine.memory(wasm);
  const versioned = engine.versionedMap(bytes("proofs"));
  await versioned.initialize();
  await versioned.put(bytes("k"), bytes("v"));
  const snapshot = await versioned.snapshot();
  assert.ok(snapshot);
  assert.equal(Buffer.from(snapshot.proveKey(bytes("k")).verify().value ?? []).toString(), "v");
  assert.equal(snapshot.stats().itemCount, 1n);
  assert.ok(snapshot.exportSummary().itemCount > 0n);
  const session = snapshot.read();
  assert.equal(Buffer.from(session.get(bytes("k")) ?? []).toString(), "v");
  assert.ok(versioned.verifyCatalog().itemCount >= 2n);
  assert.ok((await versioned.backup()).byteLength > 0);
  assert.ok(versioned.planGc().itemCount > 0n);

  const registry = engine.indexRegistry();
  registry.register({ name: bytes("by_value"), generation: 1n, extractorId: "value-v1", projection: "all",
    extract: (_key: Uint8Array, value: Uint8Array) => [{ term: Uint8Array.from(value) }] });
  const indexed = engine.indexedMap(bytes("indexed-maintenance"), registry);
  const version: any = await indexed.put(bytes("k"), bytes("term"));
  await indexed.ensureIndex(bytes("by_value"));
  assert.equal(indexed.verifyIndex(bytes("by_value"), version.sourceVersion).valid, true);
  assert.ok(indexed.metrics().buildAttempts >= 1n);
  assert.ok(indexed.exportCurrent().byteLength > 0);

  const proximity = await engine.buildProximity(2, [
    { key: bytes("p"), vector: new Float32Array([0, 0]), value: bytes("payload") },
  ]);
  assert.equal(Buffer.from(proximity.proveMembership(bytes("p")).verify(proximity.descriptor()).value ?? []).toString(), "payload");
  assert.equal(proximity.verify().recordCount, 1n);
  const searchProof = proximity.proveSearch(new Float32Array([0, 0]), 1);
  const verifiedSearch = searchProof.verify(proximity.descriptor());
  assert.equal(Buffer.from(verifiedSearch.result.neighbors[0].key).toString(), "p");
  assert.ok(verifiedSearch.replayedEvents > 0n);
  engine.close();
});

test("WASM indexed maps expose batch CAS and historical snapshots", { skip: !generatedPresent }, async () => {
  const engine = api.Engine.memory(wasm);
  const registry = engine.indexRegistry();
  registry.register({ name: bytes("by_value"), generation: 1n, extractorId: "value-v1", projection: "all",
    extract: (_key: Uint8Array, value: Uint8Array) => [{ term: Uint8Array.from(value) }] });
  const indexed = engine.indexedMap(bytes("indexed-lifecycle"), registry);
  assert.equal(Buffer.from(indexed.id()).toString(), "indexed-lifecycle");
  const first = await indexed.apply([
    { kind: "upsert", key: bytes("u1"), value: bytes("red") },
    { kind: "upsert", key: bytes("u2"), value: bytes("red") },
  ]);
  await indexed.ensureIndex(bytes("by_value"));
  const firstSnapshot = await indexed.snapshot();
  const firstId = firstSnapshot.id();
  assert.deepEqual(firstId.sourceVersion, first.sourceVersion);
  const applied = await indexed.applyIf(first.sourceVersion, [
    { kind: "upsert", key: bytes("u3"), value: bytes("blue") },
  ]);
  assert.equal(applied.kind, "applied");
  const conflict = await indexed.applyIf(first.sourceVersion, [
    { kind: "delete", key: bytes("u1") },
  ]);
  assert.equal(conflict.kind, "conflict");
  const historical = await indexed.snapshotAt(first.sourceVersion);
  assert.deepEqual(historical.id().sourceVersion, firstId.sourceVersion);
  const reopened = await indexed.snapshotById(firstId);
  assert.deepEqual(reopened.id(), firstId);
  engine.close();
});

test("WASM indexed maintenance and every bounded page direction are complete", { skip: !generatedPresent }, async () => {
  const engine = api.Engine.memory(wasm);
  const registry = engine.indexRegistry();
  registry.register({ name: bytes("by_value"), generation: 1n, extractorId: "value-v1", projection: "all",
    extract: (_key: Uint8Array, value: Uint8Array) => [{ term: Uint8Array.from(value) }] });
  const indexed = engine.indexedMap(bytes("indexed-records"), registry);
  const version = await indexed.apply([
    { kind: "upsert", key: bytes("u1"), value: bytes("red") },
    { kind: "upsert", key: bytes("u2"), value: bytes("red") },
    { kind: "upsert", key: bytes("u3"), value: bytes("rose") },
  ]);
  await indexed.ensureIndex(bytes("by_value"));
  assert.equal(Buffer.from(indexed.health().sourceMapId).toString(), "indexed-records");
  assert.equal(indexed.verifyIndex(bytes("by_value"), version.sourceVersion).valid, true);
  assert.equal(indexed.verifyAll(version.sourceVersion).length, 1);
  assert.equal(indexed.repairIndex(bytes("by_value"), version.sourceVersion).canonical, true);
  assert.ok(indexed.metrics().buildAttempts >= 1n);

  const index = (await indexed.snapshot()).index(bytes("by_value"));
  assert.equal(Buffer.from(index.name()).toString(), "by_value");
  assert.equal((await index.prefix(bytes("r"))).length, 3);
  assert.equal((await index.range(bytes("red"), bytes("s"))).length, 3);
  assert.equal(Buffer.from((await index.exactPage(bytes("red"), undefined, 1n)).matches[0].primaryKey).toString(), "u1");
  assert.equal(Buffer.from((await index.exactReversePage(bytes("red"), undefined, 1n)).matches[0].primaryKey).toString(), "u2");
  assert.equal(Buffer.from((await index.prefixPage(bytes("r"), undefined, 1n)).matches[0].primaryKey).toString(), "u1");
  assert.equal(Buffer.from((await index.prefixReversePage(bytes("r"), undefined, 1n)).matches[0].primaryKey).toString(), "u3");
  assert.equal(Buffer.from((await index.rangePage(bytes("red"), bytes("s"), undefined, 1n)).matches[0].primaryKey).toString(), "u1");
  assert.equal(Buffer.from((await index.rangeReversePage(bytes("red"), bytes("s"), undefined, 1n)).matches[0].primaryKey).toString(), "u3");

  const bundle = indexed.exportCurrent();
  const next = await indexed.put(bytes("u4"), bytes("blue"));
  assert.deepEqual((await indexed.importCurrent(bundle, next.sourceVersion)).sourceVersion, version.sourceVersion);
  assert.ok(indexed.keepLast(1n).retainedSourceVersions.length >= 1);
  await indexed.deactivateIndex(bytes("by_value"));
  assert.equal(indexed.health().activeIndexes.length, 0);
  engine.close();
});

File: bindings/go/portable_parity_test.go
package prolly

import (
	"bytes"
	"context"
	"errors"
	"testing"
)

func TestPortableVersionedIndexedAndProximityMaps(t *testing.T) {
	config, err := DefaultConfig()
	if err != nil {
		t.Fatal(err)
	}
	engine, err := NewMemoryEngine(config)
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(engine.Close)

	versioned, err := engine.VersionedMap([]byte("users"))
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(versioned.Close)
	if _, err := versioned.Initialize(); err != nil {
		t.Fatal(err)
	}
	if _, err := versioned.Put([]byte("u1"), []byte("Ada")); err != nil {
		t.Fatal(err)
	}
	value, ok, err := versioned.Get([]byte("u1"))
	if err != nil || !ok || !bytes.Equal(value, []byte("Ada")) {
		t.Fatalf("versioned get = %q, %v, %v", value, ok, err)
	}

	registry, err := NewIndexRegistry()
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(registry.Close)
	err = registry.Register(
		[]byte("by_team"), 1, "team-v1", IndexProjectionAll, nil,
		IndexExtractorFunc(func(_ []byte, source []byte) ([]IndexEntry, error) {
			return []IndexEntry{{Term: bytes.Clone(source)}}, nil
		}),
	)
	if err != nil {
		t.Fatal(err)
	}
	indexed, err := engine.IndexedMap([]byte("members"), registry)
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(indexed.Close)
	if _, err := indexed.Put([]byte("u1"), []byte("red")); err != nil {
		t.Fatal(err)
	}
	if _, err := indexed.EnsureIndex([]byte("by_team")); err != nil {
		t.Fatal(err)
	}
	snapshot, err := indexed.Snapshot()
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(snapshot.Close)
	index, err := snapshot.Index([]byte("by_team"))
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(index.Close)
	rows, err := index.Records([]byte("red"))
	if err != nil {
		t.Fatal(err)
	}
	if len(rows) != 1 || !bytes.Equal(rows[0].PrimaryKey, []byte("u1")) || !bytes.Equal(rows[0].SourceValue, []byte("red")) {
		t.Fatalf("joined index rows = %#v", rows)
	}

	var escaped IndexMatchView
	if err := index.QueryView(context.Background(), ExactIndex([]byte("red")), func(row IndexMatchView) bool {
		escaped = row
		if !bytes.Equal(row.PrimaryKey.Bytes(), []byte("u1")) {
			t.Fatalf("view primary key = %q", row.PrimaryKey.Bytes())
		}
		return true
	}); err != nil {
		t.Fatal(err)
	}
	if _, err := escaped.PrimaryKey.Copy(); !errors.Is(err, ErrViewExpired) {
		t.Fatalf("escaped view error = %v", err)
	}

	proximity, err := engine.BuildProximity(
		2,
		[]ProximityRecord{{Key: []byte("a"), Vector: []float32{0, 0}, Value: []byte("alpha")}},
	)
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(proximity.Close)
	session, err := proximity.Read()
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(session.Close)
	result, err := session.Search(context.Background(), ExactSearch([]float32{0.1, 0.1}, 1))
	if err != nil {
		t.Fatal(err)
	}
	if len(result.Neighbors) != 1 || !bytes.Equal(result.Neighbors[0].Key, []byte("a")) {
		t.Fatalf("neighbors = %#v", result.Neighbors)
	}
}

func TestPortableContextCancellation(t *testing.T) {
	engine, err := OpenMemory()
	if err != nil {
		t.Fatal(err)
	}
	defer engine.Close()
	proximity, err := engine.BuildProximity(1, []ProximityRecord{{Key: []byte("a"), Vector: []float32{0}}})
	if err != nil {
		t.Fatal(err)
	}
	defer proximity.Close()
	session, err := proximity.Read()
	if err != nil {
		t.Fatal(err)
	}
	defer session.Close()

	ctx, cancel := context.WithCancel(context.Background())
	cancel()
	_, err = session.Search(ctx, ExactSearch([]float32{0}, 1))
	if !errors.Is(err, context.Canceled) {
		t.Fatalf("search error = %v", err)
	}
}

func TestPortableAsyncWrappersCopyInputsBeforeHandoff(t *testing.T) {
	engine, err := OpenMemory()
	if err != nil {
		t.Fatal(err)
	}
	defer engine.Close()
	versioned, err := engine.VersionedMap([]byte("async"))
	if err != nil {
		t.Fatal(err)
	}
	defer versioned.Close()
	if _, err := versioned.Initialize(); err != nil {
		t.Fatal(err)
	}

	key := []byte("original-key")
	value := []byte("original-value")
	future := versioned.PutAsync(context.Background(), key, value)
	copy(key, []byte("mutated-key!"))
	copy(value, []byte("mutated-value!"))
	if _, err := future.Await(context.Background()); err != nil {
		t.Fatal(err)
	}
	got, ok, err := versioned.Get([]byte("original-key"))
	if err != nil || !ok || !bytes.Equal(got, []byte("original-value")) {
		t.Fatalf("async put get = %q, %v, %v", got, ok, err)
	}

	cancelled, cancel := context.WithCancel(context.Background())
	cancel()
	if _, err := versioned.GetAsync(cancelled, []byte("original-key")).Await(context.Background()); !errors.Is(err, context.Canceled) {
		t.Fatalf("cancelled async get error = %v", err)
	}
}

func TestPortableProofSessionAndMaintenance(t *testing.T) {
	engine, err := OpenMemory()
	if err != nil {
		t.Fatal(err)
	}
	defer engine.Close()
	versioned, err := engine.VersionedMap([]byte("proofs"))
	if err != nil {
		t.Fatal(err)
	}
	defer versioned.Close()
	if _, err = versioned.Initialize(); err != nil {
		t.Fatal(err)
	}
	if _, err = versioned.Put([]byte("k"), []byte("v")); err != nil {
		t.Fatal(err)
	}
	snapshot, err := versioned.Snapshot()
	if err != nil {
		t.Fatal(err)
	}
	defer snapshot.Close()
	proof, err := snapshot.ProveKey([]byte("k"))
	if err != nil {
		t.Fatal(err)
	}
	verified, err := VerifyKeyProof(proof)
	if err != nil || !verified.Valid || string(verified.Value) != "v" {
		t.Fatalf("proof = %+v, %v", verified, err)
	}
	session, err := snapshot.Read()
	if err != nil {
		t.Fatal(err)
	}
	defer session.Close()
	value, ok, err := session.Get([]byte("k"))
	if err != nil || !ok || string(value) != "v" {
		t.Fatalf("read = %q, %v, %v", value, ok, err)
	}
	backup, err := versioned.Backup()
	if err != nil || len(backup) == 0 {
		t.Fatalf("backup = %d, %v", len(backup), err)
	}
	catalog, err := versioned.VerifyCatalog()
	if err != nil || catalog.VersionCount < 2 {
		t.Fatalf("catalog = %+v, %v", catalog, err)
	}
}

func TestPortableIndexedProofAndMaintenance(t *testing.T) {
	engine, err := OpenMemory()
	if err != nil {
		t.Fatal(err)
	}
	defer engine.Close()
	registry, err := NewIndexRegistry()
	if err != nil {
		t.Fatal(err)
	}
	defer registry.Close()
	if err := registry.Register(
		[]byte("by_team"), 1, "team-v1", IndexProjectionAll, nil,
		IndexExtractorFunc(func(_ []byte, source []byte) ([]IndexEntry, error) {
			return []IndexEntry{{Term: bytes.Clone(source)}}, nil
		}),
	); err != nil {
		t.Fatal(err)
	}
	indexed, err := engine.IndexedMap([]byte("indexed-maintenance"), registry)
	if err != nil {
		t.Fatal(err)
	}
	defer indexed.Close()
	version, err := indexed.Put([]byte("u1"), []byte("red"))
	if err != nil {
		t.Fatal(err)
	}
	if _, err := indexed.EnsureIndex([]byte("by_team")); err != nil {
		t.Fatal(err)
	}
	verification, err := indexed.VerifyIndex([]byte("by_team"), version.SourceVersion)
	if err != nil || !verification.Valid || !verification.Canonical {
		t.Fatalf("verification = %+v, %v", verification, err)
	}
	all, err := indexed.VerifyAll(version.SourceVersion)
	if err != nil || len(all) != 1 || !all[0].Valid {
		t.Fatalf("verify all = %+v, %v", all, err)
	}
	metrics, err := indexed.Metrics()
	if err != nil || metrics.VerificationOutcomes < 2 {
		t.Fatalf("metrics = %+v, %v", metrics, err)
	}
	bundle, err := indexed.ExportCurrent()
	if err != nil || len(bundle) == 0 {
		t.Fatalf("export = %d, %v", len(bundle), err)
	}
	retention, err := indexed.KeepLast(1)
	if err != nil || len(retention.RetainedSourceVersions) != 1 {
		t.Fatalf("retention = %+v, %v", retention, err)
	}
}

func TestPortableProximityProofAndMaintenance(t *testing.T) {
	engine, err := OpenMemory()
	if err != nil {
		t.Fatal(err)
	}
	defer engine.Close()
	proximity, err := engine.BuildProximity(
		2,
		[]ProximityRecord{{Key: []byte("a"), Vector: []float32{0, 0}, Value: []byte("alpha")}},
	)
	if err != nil {
		t.Fatal(err)
	}
	defer proximity.Close()
	if count, err := proximity.Count(); err != nil || count != 1 {
		t.Fatalf("count = %d, %v", count, err)
	}
	if ok, err := proximity.Contains([]byte("a")); err != nil || !ok {
		t.Fatalf("contains = %v, %v", ok, err)
	}
	record, ok, err := proximity.Get([]byte("a"))
	if err != nil || !ok || !bytes.Equal(record.Value, []byte("alpha")) {
		t.Fatalf("record = %+v, %v, %v", record, ok, err)
	}
	descriptor, err := proximity.Descriptor()
	if err != nil || len(descriptor) == 0 {
		t.Fatalf("descriptor = %x, %v", descriptor, err)
	}
	proof, err := proximity.ProveMembership([]byte("a"))
	if err != nil {
		t.Fatal(err)
	}
	verified, err := VerifyProximityMembershipProof(proof, descriptor)
	if err != nil || !bytes.Equal(verified.Key, []byte("a")) || verified.Record == nil {
		t.Fatalf("verified = %+v, %v", verified, err)
	}
	searchProof, err := proximity.ProveSearch(ExactSearch([]float32{0, 0}, 1))
	if err != nil {
		t.Fatal(err)
	}
	defer searchProof.Close()
	verifiedSearch, err := searchProof.Verify(descriptor)
	if err != nil || len(verifiedSearch.Result.Neighbors) != 1 ||
		!bytes.Equal(verifiedSearch.Result.Neighbors[0].Key, []byte("a")) ||
		verifiedSearch.ReplayedEvents == 0 {
		t.Fatalf("verified search = %+v, %v", verifiedSearch, err)
	}
	health, err := proximity.Verify()
	if err != nil || health.RecordCount != 1 {
		t.Fatalf("health = %+v, %v", health, err)
	}
	if err := proximity.ClearContentCache(); err != nil {
		t.Fatal(err)
	}
}
